### PR TITLE
HUM-50 Bulk upload / delete middleware

### PR DESCRIPTION
**Pull request by gholt**

### CODE REVIEW

File: proxyserver/middleware/context.go [hunk 1]
```diff
@@ -117,16 +117,24 @@ func (ctx *ProxyContext) addSubrequestCopy(f subrequestCopy) {
...
+		return parts[1], parts[2], parts[3], parts[4]
+	case 4:
+		return parts[1], parts[2], parts[3], ""
+	case 3:
+		return parts[1], parts[2], "", ""
+	case 2:
+		return parts[1], "", "", ""
+	default:
+		return "", "", "", ""
 	}
```
Comment (gholt) [inline proxyserver/middleware/context.go]: Well, I was fine with the check and it's still there with the original getPathParts. But getPathParts returns a bool for that part and I need the actual string to construct more paths in bulk.

File: proxyserver/middleware/bulk.go [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+		j := &js{
+			ResponseStatus:     fmt.Sprintf("%d %s", responseStatus, http.StatusText(responseStatus)),
+			ResponseBody:       responseBody,
+			NumberFilesCreated: numberFilesCreated,
+			Errors:             [][]string{},
+		}
+		j.Errors = failures
+		b, err := json.Marshal(j)
+		if err != nil {
+			writer.Write([]byte(fmt.Sprintf("JSON encoding error: %s\n%#v\n", err, j)))
```
Comment (gholt) [inline proxyserver/middleware/bulk.go]: Yeah, I can switch that up and return a 5xx and log the json err.

File: proxyserver/middleware/bulk.go [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+		x.ResponseBody = responseBody
+		x.NumberFilesCreated = numberFilesCreated
+		for _, failure := range failures {
+			x.Error.Object = append(x.Error.Object, &errorObject{failure[0], failure[1]})
+		}
+		b, err := xml.Marshal(x)
+		if err != nil {
+			writer.Write([]byte(fmt.Sprintf("XML encoding error: %s\n%#v\n", err, x)))
+		}
+		// Yes, even the PUTs are labeled <delete> in the XML from Swift's code.
```
Comment (gholt) [inline proxyserver/middleware/bulk.go]: Well, probably, but it's Swift behavior so...

File: proxyserver/middleware/bulk_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
...1a`F\x11\x80\x00\x00\t\x84`\x15DCSF\x84\x8ci\xa84\x00i\x91\xea4\xd0\x001\xa9\xbf\xf2'\xed\xd3\xf8\xd5Z\xeb\u007fM\x8e\xeb4dP\xa3\x82\x12D\xe5\x10\x88\x8e-\x9bp\u0676o\xe03\x96$`EK-Y\x99R\xd6\u365a^1\x81\x12\x00\xa6\x9e\x94%\xe9iZ2AQ\xdb\x10Tn3\xa9\x02\xb4\u715e=\xb3\xf2\xe6\u00e2\x15\xbf\xa5\x1d\xe8h\x87\u03d1\xaa_\x14\xf2\xecL\u0489\x9aL]\xe5p\x98\xc1$\xa4\x1bH\xe2,\u0538\x0e\x9d\u06fa99\xfa\xba\xb4\x1c\u0108\xce2\xd4EH\xa8\xf1\xd6d\xf1G\x05\"\xe8F\x12m\xbb\bRy\x9d\u063d\x18\x8eh\xa5\xa0\x9b\x8er$\xe4f\xe7\xc9V\x02\xcc)H\xbf\xd1$\xe1[-z\xd4+\x8b\xea\xa8\x1e\x00s\n\x11\x90\x06x\xe54\xb5\x86\x04\xc8\x19\u0455\xe0\xc3B\xa3\x0f\x9b\x19\xd3V\"ZW\x0e\xfbR\"\xcb\u023aI\x8b\xd9\xe9\x05Vcy\x95\xd7\xe6`\xd16/\xa1\xc0\u0366\u0297/a\x12\xbc\xc64\xda%6R2\x16\xcd1\x936#\x04\xb0V\u9f85X\x18\x19S\b\xacgX\x8b\xe7\x01\xe3#\x96t:\x9fR\x9ew\xac\xa6\xe6\xe3\xa5\u07bb\xe2\x9dm\xb9\x89IE*\xf3'L\x8fM!\xfc\xec\xa8\xc0\xc9\xf4;\n\x9d\x10\xf8+9\x9drJ\x17)?l\a\xb1\xf8\x86\xa9\xc4\v\x1f\xc0o\x13\x02}]\xe1B\x9c\x04\xdb\a!3\xb8S\xe5\x1a#K\u007f~\xfcF\u05a5b\x90\xf2\x9b\x1a2\xa2\x1f\x03\xd0v\x8d/\x064\x1eN>\xd9\xd7JL\xcd#Z]\x9e\xa3\x97[\xaf~\xf1\u00e4o\x8a\x19\xcb&\xa7\x8d$\xbf\xc5\u0711N\x14$\x1d\x84\x03T\xc0")
```
Comment (gholt) [inline proxyserver/middleware/bulk_test.go]: Haha, go-bindata btw, if you ever need to embed "binaries" in your Go code.

**Comment by dpgoetz** [comment]
lgtm but will wait for @corystone

**Pull request merged**